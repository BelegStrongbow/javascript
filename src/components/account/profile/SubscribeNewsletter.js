import PropTypes from "prop-types";
import React from "react";
import { injectIntl, intlShape, FormattedMessage, FormattedHTMLMessage, defineMessages } from "react-intl";
import styled from "styled-components";

// Custom components
import { Paper, Page } from "../../PaperStyles";
import { Button, LargeSecondaryButton } from "../../Button";
import ErrorDisplay from "../../../errors/ErrorDisplay";

const NewsletterContainer = styled.div`
	margin-top: 24px;
`;

const NewsletterSection = styled.section`
	margin-bottom: 1em;
`;

const Paragraph = styled.p`
	margin-top: 0.0em;
	margin-bottom: 0.5em;
	font-size: 1.5em;
	font-weight: 300;
`;

const messages = defineMessages( {
	title: {
		id: "newsletter.title",
		defaultMessage: "Newsletter",
	},
	subscribeReasons: {
		id: "newsletter.subscribe.reasons",
		defaultMessage: "Subscribe to get weekly tips on optimizing your website's SEO, usability and conversion.<br>" +
		"You'll also be the first to know about new features and other cool (free) plugins!",
	},
	subscribeButton: {
		id: "newsletter.subscribeButton",
		defaultMessage: "Subscribe",
	},
	unsubscribeParagraph: {
		id: "newsletter.unsubscribeParagraph",
		defaultMessage: "You are subscribed to the Yoast newsletter. If you want to unsubscribe, click the button below.",
	},
	pendingParagraph: {
		id: "newsletter.pendingParagraph",
		defaultMessage: "Your subscription to the Yoast newsletter is pending. Please check your inbox for a confirmation email.",
	},
	unsubscribeButton: {
		id: "newsletter.unSubscribeButton",
		defaultMessage: "Unsubscribe",
	},
	privacyPolicy: {
		id: "newsletter.privacyPolicy",
		defaultMessage: "Yoast respects your privacy. Read our <a href='https://yoast.com/privacy-policy/'>privacy policy</a> on how we handle your personal information.",
	},
} );

class SubscribeNewsletter extends React.Component {
	/**
	 * Constructor
	 *
	 * @param {Object} props The props.
	 */
	constructor( props ) {
		super( props );

		this.onClick = this.onClick.bind( this );
	}

	/**
	 * Handles clicks on the button.
	 * Nothing will happen when loading.
	 * A user will be unsubscribed when subscribed.
	 * Otherwise a user will be subscribed.
	 *
	 * @returns {void}
	 */
	onClick() {
		if ( this.props.loading ) {
			return;
		}

		if ( this.props.subscribed === "subscribed" ) {
			this.props.onUnsubscribe();
			return;
		}

		this.props.onSubscribe();
	}

	/**
	 * Generates and returns the (un)subscribe paragraph,
	 * based on the current subscription status of the customer.
	 *
	 * E.g. invites customer to subscribe when not yet subscribed,
	 * and explains how to unsubscribe when subscribed.
	 *
	 * @returns {Element} an element containing the content
	 */
	getContent() {
		if ( this.props.subscribed === "subscribed" ) {
			return (
				<p>
					<FormattedMessage
					id={ messages.unsubscribeParagraph.id }
					defaultMessage={ messages.unsubscribeParagraph.defaultMessage }/>
				</p>
			);
		}

		if ( this.props.subscribed === "pending" ) {
			return (
				<p>
					<FormattedMessage
						id={ messages.pendingParagraph.id }
						defaultMessage={ messages.pendingParagraph.defaultMessage }/>
				</p>
			);
		}

		return (
			<p>
				<FormattedHTMLMessage
					id={ messages.subscribeReasons.id }
					defaultMessage={ messages.subscribeReasons.defaultMessage }/>
			</p>
		);
	}

	/**
	 * Generates and returns the (un)subscribe button,
	 * based on the current subscription status of the customer.
	 *
	 * @returns {Button} a button to subscribe or unsubscribe
	 */
	getButton() {
		if ( this.props.subscribed === "subscribed" ) {
			return (
				<LargeSecondaryButton onClick={ this.onClick }>
					<FormattedMessage id={ messages.unsubscribeButton.id } defaultMessage={ messages.unsubscribeButton.defaultMessage }/>
				</LargeSecondaryButton>
			);
		}

		if ( this.props.subscribed === "pending" ) {
			return null;
		}

		return (
			<Button onClick={ this.onClick }>
				<FormattedMessage id={ messages.subscribeButton.id } defaultMessage={ messages.subscribeButton.defaultMessage }/>
			</Button>
		);
	}

	render() {
		if ( this.props.subscribed === "unknown" ) {
			return null;
		}

		return (
			<NewsletterContainer>
				<Paper>
					<Page>
						<NewsletterSection>
							<Paragraph>
								<FormattedMessage id={ messages.title.id } defaultMessage={ messages.title.defaultMessage }/>
							</Paragraph>
							{ this.getContent() }
							<p>
								<FormattedHTMLMessage id={ messages.privacyPolicy.id } defaultMessage={ messages.privacyPolicy.defaultMessage }/>
							</p>
							{ this.getButton() }
							{ this.props.error &&
								<ErrorDisplay error={ { code: "MAILCHIMP_ERROR", message: this.props.error } } />
							}
						</NewsletterSection>
					</Page>
				</Paper>
			</NewsletterContainer>
		);
	}
}

SubscribeNewsletter.propTypes = {
	intl: intlShape.isRequired,
	onSubscribe: PropTypes.func.isRequired,
	onUnsubscribe: PropTypes.func.isRequired,
	loading: PropTypes.bool,
	subscribed: PropTypes.string,
	error: PropTypes.string,
};


export default injectIntl( SubscribeNewsletter );
