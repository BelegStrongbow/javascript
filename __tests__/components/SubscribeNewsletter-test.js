import React from 'react';
import { createComponentWithIntl } from "../../utils";
import SubscribeNewsletter from '../../src/components/account/profile/SubscribeNewsletter';

test('The Subscribe Newsletter component (when subscribed) matches the snapshot', () => {
	const component = createComponentWithIntl(
		<SubscribeNewsletter
			onSubscribe={ () => null }
			onUnsubscribe={ () => null }
			subscribed={ "subscribed" }
			loading={ false }
			error={ null }
		/>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('The Subscribe Newsletter component (when unsubscribed) matches the snapshot', () => {
	const component = createComponentWithIntl(
		<SubscribeNewsletter
			onSubscribe={ () => null }
			onUnsubscribe={ () => null }
			subscribed={ "unsubscribed" }
			loading={ false }
			error={ null }
		/>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('The Subscribe Newsletter component (when pending) matches the snapshot', () => {
	const component = createComponentWithIntl(
		<SubscribeNewsletter
			onSubscribe={ () => null }
			onUnsubscribe={ () => null }
			subscribed={ "pending" }
			loading={ false }
			error={ null }
		/>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('The Subscribe Newsletter component (when unknown) matches the snapshot', () => {
	const component = createComponentWithIntl(
		<SubscribeNewsletter
			onSubscribe={ () => null }
			onUnsubscribe={ () => null }
			subscribed={ "unknown" }
			loading={ false }
			error={ null }
		/>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('The Subscribe Newsletter component shows an error when there is one defined.', () => {
	const component = createComponentWithIntl(
		<SubscribeNewsletter
			onSubscribe={ () => null }
			onUnsubscribe={ () => null }
			subscribed={ "unsubscribed" }
			loading={ false }
			error={ "An error occurred" }
		/>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('The Subscribe Newsletter component subscribes the customer when he/she clicks on the button and is unsubscribed.', () => {

	const mockupFunction = jest.fn();

	const component = createComponentWithIntl(
		<SubscribeNewsletter
			onSubscribe={ mockupFunction }
			onUnsubscribe={ () => null }
			subscribed={ "unsubscribed" }
			loading={ false }
			error={ null }
		/>
	);

	let tree = component.toJSON();

	// call onClick function on the component.
	const onClick = tree.children[2].props.onClick;
	onClick();

	expect(mockupFunction).toHaveBeenCalled();
});

test('The Subscribe Newsletter component unsubscribes the customer when he/she clicks on the button and is subscribed.', () => {

	const mockupFunction = jest.fn();

	const component = createComponentWithIntl(
		<SubscribeNewsletter
			onSubscribe={ () => null }
			onUnsubscribe={ mockupFunction }
			subscribed={ "subscribed" }
			loading={ false }
			error={ null }
		/>
	);

	let tree = component.toJSON();

	// call onClick function on the component.
	const onClick = tree.children[2].props.onClick;
	onClick();

	expect(mockupFunction).toHaveBeenCalled();
});

test('The Subscribe Newsletter component button does nothing when the component is loading.', () => {

	const mockupFunction = jest.fn();

	const component = createComponentWithIntl(
		<SubscribeNewsletter
			onSubscribe={ mockupFunction }
			onUnsubscribe={ mockupFunction }
			subscribed={ "subscribed" }
			loading={ true }
			error={ null }
		/>
	);

	let tree = component.toJSON();

	console.error( tree );

	// call onClick function on the component.
	const onClick = tree.children[2].props.onClick;
	onClick();

	expect(mockupFunction).not.toHaveBeenCalled();
});
