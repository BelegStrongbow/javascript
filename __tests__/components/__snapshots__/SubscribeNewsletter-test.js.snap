// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Subscribe Newsletter component (when pending) matches the snapshot 1`] = `
.c1 {
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
}

.c3 {
  background-color: #fff;
  padding: 16px 24px;
}

.c2 {
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  margin-top: 24px;
}

.c4 {
  margin-bottom: 1em;
}

.c5 {
  margin-top: 0.0em;
  margin-bottom: 0.5em;
  font-size: 1.5em;
  font-weight: 300;
}

@media screen and ( max-width:800px ) {
  .c2 {
    display: block;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2 c3"
    >
      <section
        className="c4"
      >
        <p
          className="c5"
        >
          <span>
            Newsletter
          </span>
        </p>
        <p>
          <span>
            Your subscription to the Yoast newsletter is pending. Please check your inbox for a confirmation email.
          </span>
        </p>
        <p>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Yoast respects your privacy. Read our <a href='https://yoast.com/privacy-policy/'>privacy policy</a> on how we handle your personal information.",
              }
            }
          />
        </p>
      </section>
    </div>
  </div>
</div>
`;

exports[`The Subscribe Newsletter component (when subscribed) matches the snapshot 1`] = `
.c1 {
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
}

.c3 {
  background-color: #fff;
  padding: 16px 24px;
}

.c2 {
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c7 {
  height: 48px;
  margin: 0;
  padding: 0 16px;
  border: 0;
  border-bottom: 2px solid rgba(0,0,0,0.2);
  background-color: #64a60a;
  color: #fff;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
  border-radius: 4px;
  font: 400 14px/24px "Open Sans",sans-serif;
  text-transform: uppercase;
  cursor: pointer;
  vertical-align: top;
  -webkit-transition: background-color 150ms ease-out;
  transition: background-color 150ms ease-out;
}

.c7:hover,
.c7:focus {
  box-shadow: 0 4px 10px 0 rgba(0,0,0,0.2),inset 0 0 0 100px rgba( 0,0,0,0.1 );
}

.c7:active {
  -webkit-transform: translateY( 1px );
  -ms-transform: translateY( 1px );
  transform: translateY( 1px );
  box-shadow: none;
}

.c6 {
  color: #000;
  box-shadow: none;
  background-color: #fff;
  border: 1px solid #000;
  min-width: 152px;
}

.c6:hover,
.c6:focus {
  background-color: #64a60a;
  color: #fff;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
  border: 0;
  text-shadow: 0px 0px 2px #000;
}

.c0 {
  margin-top: 24px;
}

.c4 {
  margin-bottom: 1em;
}

.c5 {
  margin-top: 0.0em;
  margin-bottom: 0.5em;
  font-size: 1.5em;
  font-weight: 300;
}

@media screen and ( max-width:800px ) {
  .c2 {
    display: block;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2 c3"
    >
      <section
        className="c4"
      >
        <p
          className="c5"
        >
          <span>
            Newsletter
          </span>
        </p>
        <p>
          <span>
            You are subscribed to the Yoast newsletter. If you want to unsubscribe, click the button below.
          </span>
        </p>
        <p>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Yoast respects your privacy. Read our <a href='https://yoast.com/privacy-policy/'>privacy policy</a> on how we handle your personal information.",
              }
            }
          />
        </p>
        <button
          className="c6 c7"
          onClick={[Function]}
          type="button"
        >
          <span>
            Unsubscribe
          </span>
        </button>
      </section>
    </div>
  </div>
</div>
`;

exports[`The Subscribe Newsletter component (when unknown) matches the snapshot 1`] = `null`;

exports[`The Subscribe Newsletter component (when unsubscribed) matches the snapshot 1`] = `
.c1 {
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
}

.c3 {
  background-color: #fff;
  padding: 16px 24px;
}

.c2 {
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c7 {
  height: 48px;
  margin: 0;
  padding: 0 16px;
  border: 0;
  border-bottom: 2px solid rgba(0,0,0,0.2);
  background-color: #64a60a;
  color: #fff;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
  border-radius: 4px;
  font: 400 14px/24px "Open Sans",sans-serif;
  text-transform: uppercase;
  cursor: pointer;
  vertical-align: top;
  -webkit-transition: background-color 150ms ease-out;
  transition: background-color 150ms ease-out;
}

.c7:hover,
.c7:focus {
  box-shadow: 0 4px 10px 0 rgba(0,0,0,0.2),inset 0 0 0 100px rgba( 0,0,0,0.1 );
}

.c7:active {
  -webkit-transform: translateY( 1px );
  -ms-transform: translateY( 1px );
  transform: translateY( 1px );
  box-shadow: none;
}

.c6 {
  text-shadow: 0px 0px 2px #000;
}

.c0 {
  margin-top: 24px;
}

.c4 {
  margin-bottom: 1em;
}

.c5 {
  margin-top: 0.0em;
  margin-bottom: 0.5em;
  font-size: 1.5em;
  font-weight: 300;
}

@media screen and ( max-width:800px ) {
  .c2 {
    display: block;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2 c3"
    >
      <section
        className="c4"
      >
        <p
          className="c5"
        >
          <span>
            Newsletter
          </span>
        </p>
        <p>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Subscribe to get weekly tips on optimizing your website's SEO, usability and conversion.<br>You'll also be the first to know about new features and other cool (free) plugins!",
              }
            }
          />
        </p>
        <p>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Yoast respects your privacy. Read our <a href='https://yoast.com/privacy-policy/'>privacy policy</a> on how we handle your personal information.",
              }
            }
          />
        </p>
        <button
          className="c6 c7"
          onClick={[Function]}
          type="button"
        >
          <span>
            Subscribe
          </span>
        </button>
      </section>
    </div>
  </div>
</div>
`;

exports[`The Subscribe Newsletter component shows an error when there is one defined. 1`] = `
.c1 {
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
}

.c3 {
  background-color: #fff;
  padding: 16px 24px;
}

.c2 {
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c7 {
  height: 48px;
  margin: 0;
  padding: 0 16px;
  border: 0;
  border-bottom: 2px solid rgba(0,0,0,0.2);
  background-color: #64a60a;
  color: #fff;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
  border-radius: 4px;
  font: 400 14px/24px "Open Sans",sans-serif;
  text-transform: uppercase;
  cursor: pointer;
  vertical-align: top;
  -webkit-transition: background-color 150ms ease-out;
  transition: background-color 150ms ease-out;
}

.c7:hover,
.c7:focus {
  box-shadow: 0 4px 10px 0 rgba(0,0,0,0.2),inset 0 0 0 100px rgba( 0,0,0,0.1 );
}

.c7:active {
  -webkit-transform: translateY( 1px );
  -ms-transform: translateY( 1px );
  transform: translateY( 1px );
  box-shadow: none;
}

.c6 {
  text-shadow: 0px 0px 2px #000;
}

.c9 {
  padding: 1em;
  background-color: #ffeb3b;
  margin: 0.5em 0 0 0;
  overflow: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c8 {
  background-color: #f9bdbd;
}

.c10 {
  background-position: 50%;
  background-repeat: no-repeat;
  background-size: 2em;
  background-image: url( exclamation-circle.svg );
  height: 2em;
  min-width: 2em;
  margin: 0 1em;
}

.c11 {
  color: #5d237a;
}

.c0 {
  margin-top: 24px;
}

.c4 {
  margin-bottom: 1em;
}

.c5 {
  margin-top: 0.0em;
  margin-bottom: 0.5em;
  font-size: 1.5em;
  font-weight: 300;
}

@media screen and ( max-width:800px ) {
  .c2 {
    display: block;
  }
}

@media screen and ( max-width:800 ) {
  .c9 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    text-align: left;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2 c3"
    >
      <section
        className="c4"
      >
        <p
          className="c5"
        >
          <span>
            Newsletter
          </span>
        </p>
        <p>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Subscribe to get weekly tips on optimizing your website's SEO, usability and conversion.<br>You'll also be the first to know about new features and other cool (free) plugins!",
              }
            }
          />
        </p>
        <p>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Yoast respects your privacy. Read our <a href='https://yoast.com/privacy-policy/'>privacy policy</a> on how we handle your personal information.",
              }
            }
          />
        </p>
        <button
          className="c6 c7"
          onClick={[Function]}
          type="button"
        >
          <span>
            Subscribe
          </span>
        </button>
        <div
          className="c8 c9"
          role="alert"
        >
          <span
            alt="Error sign"
            className="c10"
          />
          <span>
            Oops, something went wrong on our end, please try again. If this keeps happening, 
            <a
              className="c11"
              href="mailto:support@yoast.com"
            >
              <span>
                please contact support
              </span>
            </a>
            .
          </span>
        </div>
      </section>
    </div>
  </div>
</div>
`;
